### PR TITLE
perf(guided-decoding): optimize with async D2H copy and xgrammar v0.2.1

### PR DESCRIPTION
**Pull request by windreamer**
## Motivation

Improve guided decoding performance in TurboMind by reducing synchronization overhead and upgrading to xgrammar v0.2.1 for API compatibility.

## Modification

This PR includes the following optimizations and fixes:

### Performance Optimizations

1. **Async D2H Copy Overlap** (main optimization):
   - Split `GuidedDecoding::Update()` into `ScheduleUpdate()` + `FinishUpdate()` to enable D2H copy overlap with GPU work
   - Added secondary CUDA stream (`d2h_stream_`) for asynchronous output_ids D2H transfer
   - Added CUDA events (`sampling_done_`, `d2h_done_`) for stream synchronization
   - D2H copy now overlaps with `AppendTokenIds` and `stop_criteria` GPU kernels on the main stream
   - Eliminates a blocking `cudaStreamSynchronize` in the decode step hot path

2. **Grammar Compiler Caching**:
   - Lazy-initialized `GrammarCompiler` shared across all requests in `TurboMindModel`
   - Avoids recreating compiler + tokenizer info on every request

### Correctness Fixes

3. **Limit Iteration to Active Generation Slots**:
   - Limited iteration in `FillMask`, `ScheduleUpdate`, and `FinishUpdate` to `generation_size` (= `logits.shape(0)`)
   - Previous code iterated over all `matchers.size()` entries, including idle/prefill slots with stale `output_ids`
   - Prevents processing stale data beyond active generation slots

### Dependency Updates

4. **xgrammar v0.2.1 Upgrade**:
   - Updated from v0.1.27 to v0.2.1 in `CMakeLists.txt`
   - Added required `debug_print=False` argument to `accept_token()` API
   - Ensured token conversion to plain Python `int` before passing to `accept_token()`

### Build System

5. **CMake Dependency Cleanup**:
   - Changed `xgrammar` and `core` from `PRIVATE` to `PUBLIC` linkage in `guided_decoding` library
   - Ensures proper transitive dependency resolution for downstream targets

## BC-breaking (Optional)

No backward compatibility issues. The API changes are internal to the TurboMind engine and PyTorch engine's guided decoding implementation.

## Use cases (Optional)

This optimization benefits all users of structured generation features:
- JSON schema constrained generation
- Regex pattern constrained generation  
- JSON object generation

Example usage:
```python
from lmdeploy import pipeline, GenerationConfig

pipe = pipeline('Qwen/Qwen2.5-7B-Instruct')
response = pipe(
    'Generate a user profile',
    gen_config=GenerationConfig(
        response_format={'type': 'json_schema', 'json_schema': {...}}
    )
)
```

## Checklist

1. ✅ Pre-commit or other linting tools are used to fix the potential lint issues.
2. ✅ The modification is covered by complete unit tests. All tests pass:
   - Qwen/Qwen3-0.6B: all tests pass
   - Qwen/Qwen3-VL-2B-Instruct: all tests pass (JSON schema, regex schema, json_object)
3. ✅ No dependency on downstream projects with version changes.
4. ✅ Documentation not required for internal optimization changes.

## Test Results

All guided decoding tests pass with the new implementation:
- JSON schema constrained generation: ✅
- Regex pattern constrained generation: ✅
- JSON object generation: ✅

### CODE REVIEW

File: src/turbomind/generation/guided_decoding.cc [hunk 1]
```diff
@@ -56,24 +61,36 @@ void GuidedDecoding::FillMask(int phase, TensorMap& env)
...
+        d.needs_apply = false;
+
         if (tp_group_->rank() == 0) {
             for (size_t i = 0; i < d.matchers.size(); ++i) {
-                if (const auto& matcher = d.matchers[i]; matcher && !matcher->IsTerminated()) {
-                    matcher->FillNextTokenBitmask(&dlbitmask, i);
+                if (const auto& m = d.matchers[i]; m && !m->IsTerminated()) {
+                    active_matchers.push_back(*m);
+                    active_indices.push_back(static_cast<int32_t>(i));
                 }
```
[the remark this answers is not in the document]
Comment (windreamer) [reply src/turbomind/generation/guided_decoding.cc]: Fixed in 57050fb2 — removed `needs_apply` entirely and reverted `ApplyMask` to gate on `d.active` only. Since `d.active` is set consistently on all TP ranks in `Setup`, the broadcast now works correctly in multi-rank scenarios.

File: src/turbomind/generation/guided_decoding.cc [hunk 2]
```diff
@@ -56,24 +61,36 @@ void GuidedDecoding::FillMask(int phase, TensorMap& env)
                            (int64_t*)bitmask_buf_.shape().data(),
                            nullptr,
                            0};
+
+        std::vector<xgrammar::GrammarMatcher> active_matchers;
+        std::vector<int32_t>                  active_indices;
```
[the remark this answers is not in the document]
Comment (windreamer) [reply src/turbomind/generation/guided_decoding.cc]: Fixed in 57050fb2 — added `.reserve(d.matchers.size())` and switched to `emplace_back` for all temporary vectors in both `FillMask` and `FinishUpdate`.

File: src/turbomind/generation/guided_decoding.cc [hunk 3]
```diff
@@ -88,18 +103,48 @@ void GuidedDecoding::ApplyMask(int phase, TensorMap& env)
...
+        // Record event on main stream after sampling GPU work is submitted.
+        // The secondary stream will wait for this before issuing the D2H copy,
+        // ensuring it reads the output_ids written by sampling.
+        sampling_done_.Record(core::Context::stream());
+
+        // D2H copy on secondary stream — overlaps with subsequent GPU kernels
+        // on the main stream (AppendTokenIds, stop_criteria).
+        d2h_stream_.Wait(sampling_done_);
+        Copy(env.at("output_ids").buffer(), d.matchers.size(), output_ids_buf_, d2h_stream_);
+        d2h_done_.Record(d2h_stream_);
```
[the remark this answers is not in the document]
Comment (windreamer) [reply src/turbomind/generation/guided_decoding.cc:119]: Fixed in b5a36782 — guarded `ScheduleUpdate` and `FinishUpdate` with `tp_group_->rank() == 0` so that non-zero ranks skip the D2H copy, event recording, and sync entirely.

File: lmdeploy/pytorch/engine/logits_process.py [hunk 4]
```diff
@@ -480,7 +480,7 @@ def __random_sampling(scores: torch.Tensor, indices: torch.LongTensor):
 
         if self.guided_decoding_manager and self.guided_processors:
             for i, processor in self.guided_processors.items():
-                self.guided_decoding_manager.accept_token(processor, result[i])
+                self.guided_decoding_manager.accept_token(processor, result[i].item())
```
[the remark this answers is not in the document]
Comment (windreamer) [reply lmdeploy/pytorch/engine/logits_process.py]: Fixed in b5a36782 — replaced the per-element `.item()` loop with a single `result.tolist()` call, then iterate over the CPU list. This eliminates the repeated CUDA sync per token.

File: src/turbomind/generation/guided_decoding.cc [hunk 5]
```diff
@@ -88,18 +103,45 @@ void GuidedDecoding::ApplyMask(int phase, TensorMap& env)
...
```
[the remark this answers is not in the document]
Comment (windreamer) [reply src/turbomind/generation/guided_decoding.cc:119]: Fixed in 19646d8b — `ScheduleUpdate` now reads `gs = env.at("logits").shape(0)` and copies only `gs` entries instead of `d.matchers.size()`. No `generation_size` member was added to `Data`; the value is read directly from the logits tensor shape each time it's needed.

File: src/turbomind/generation/guided_decoding.cc [hunk 6]
```diff
@@ -56,17 +60,28 @@ void GuidedDecoding::FillMask(int phase, TensorMap& env)
...
+        active_indices.reserve(d.matchers.size());
+
         if (tp_group_->rank() == 0) {
             for (size_t i = 0; i < d.matchers.size(); ++i) {
-                if (const auto& matcher = d.matchers[i]; matcher && !matcher->IsTerminated()) {
-                    matcher->FillNextTokenBitmask(&dlbitmask, i);
+                if (const auto& m = d.matchers[i]; m && !m->IsTerminated()) {
+                    active_matchers.emplace_back(*m);
+                    active_indices.emplace_back(static_cast<int32_t>(i));
                 }
```
[the remark this answers is not in the document]
Comment (windreamer) [reply src/turbomind/generation/guided_decoding.cc]: Fixed in 19646d8b — `FillMask` now limits iteration to `gs = env.at("logits").shape(0)` instead of `d.matchers.size()`. `FinishUpdate` also received the same fix (with a new `TensorMap\& env` parameter to access the logits shape). No member variable added to `Data` — the value is read from the logits tensor shape directly.

File: src/turbomind/generation/guided_decoding.cc [hunk 7]
```diff
@@ -56,17 +65,28 @@ void GuidedDecoding::FillMask(int phase, TensorMap& env)
                            (int64_t*)bitmask_buf_.shape().data(),
                            nullptr,
                            0};
+
+        std::vector<xgrammar::GrammarMatcher> active_matchers;
+        std::vector<int32_t>                  active_indices;
+        active_matchers.reserve(gs);
+        active_indices.reserve(gs);
+
         if (tp_group_->rank() == 0) {
```
[the remark this answers is not in the document]
Comment (windreamer) [reply src/turbomind/generation/guided_decoding.cc]: Fixed in ea5c50ba — moved `active_matchers`/`active_indices` vector declarations and `.reserve()` calls inside the `if (tp_group_->rank() == 0)` block. Non-zero ranks no longer allocate these vectors in `FillMask()`.

File: src/turbomind/generation/CMakeLists.txt [hunk 8]
```diff
@@ -15,7 +15,7 @@
 cmake_minimum_required(VERSION 3.11)
 
 add_library(guided_decoding STATIC guided_decoding.cc)
-target_link_libraries(guided_decoding PRIVATE
+target_link_libraries(guided_decoding PUBLIC
     apply_token_bitmask_inplace_cuda
     xgrammar
     core)
```
[the remark this answers is not in the document]
Comment (windreamer) [reply src/turbomind/generation/CMakeLists.txt]: Fixed in ea5c50ba — split `target_link_libraries` into PUBLIC (xgrammar, core — needed by guided_decoding.h) and PRIVATE (apply_token_bitmask_inplace_cuda — only used in the .cc implementation). This avoids propagating internal-only link dependencies to downstream targets.

File: src/turbomind/generation/guided_decoding.cc [hunk 9]
```diff
@@ -25,6 +25,10 @@ GuidedDecoding::GuidedDecoding(const BaseGenerationParam& base, const comm::Host
     bitmask_buf_    = {{max_batch_size_, bitmask_size}, kCPUpinned};
     output_ids_buf_ = {max_batch_size_, kCPUpinned};
 
+    d2h_stream_    = core::Stream::create();
+    sampling_done_ = core::Event::create();
+    d2h_done_      = core::Event::create();
+
     for (int i = 0; i < phases; ++i) {
         auto& d    = data_.emplace_back(std::make_shared<Data>());
         d->bitmask = empty_like(bitmask_buf_);
```
Comment (irexyc) [inline src/turbomind/generation/guided_decoding.cc]: Why use `kCPUpinned` memory instead of `kDEVICE` memory for d->bitmask

Comment (windreamer) [reply src/turbomind/generation/guided_decoding.cc]: fixed in 87c639c

**Review by irexyc — APPROVED** [review]

**Review by lzhangzz — APPROVED** [review]
